Core_templates snapshot v1 (#1014)
openmsupply/application-manager-web-app: docs/internal/admin/List-Core-Actions.md

# List of Core Actions

> The snapshot loaded with the system at the start called `core_templates` has a list of all core-actions included in template `Sample template`.

Reviewer: ```suggestion
The snapshot loaded with the system at the start, called `core_templates`, has a template called `Sample template`, which contains all the core actions listed here.
```

## Disclaimer

The plugin actions that will be listed here are different from actions that users can act in the sytem, this are only meant for admins and configure how the system will work.
Core-actions aren't meant to be modified, and only someone with expertise should be dealing with these. Otherwise some of the features that rely on certain actions would stop working.

## Briefing

Core-actions are listed in each template and are configurations that run after certain events happen triggered by changes done by users in Applications linked to the template.

> These actions should be added at the start of a template creation - one siimple way is to duplicate the `Sample template` to start with all of these core-actions in the new template configuration.

Reviewer: ```suggestion
These actions should be added at the start of a template creation - one simple way is to duplicate the `Sample template` to start with all of these core-actions in the new template configuration, although it is recommended you duplicate one of the other core templates that mostly closely matches your requirements, and it will also have all the core actions included (with some exceptions for a couple of the simpler templates)
```

> Some of the plugin actions - which aren't core-actions - are also defined per template (and some examples at the end) used mainly for configuring notifications, generatioin of certificates and outcomes.

Reviewer: ```suggestion
Some of the plugin actions - which aren't core-actions - are also defined per template (and some examples at the end) used mainly for configuring notifications, generation of certificates, and outcomes.
```

## Lists

This list should reflect the same as we have in `Sample templates` so when changing that we need to also add to this list.

### Templates

- Edit User Details (No review)
- User Registration (No review)
- Company Registration (1 stage)
  - Review - permission: `reviewGeneral`
- Join Company (1 stage)
  - Review: - permissioin: `reviewGeneral`
- Demo -- Features Showcase (3 stages)
  - Screening - permission: `reviewScreening` self-assignable
  - Assemessment: (2 levels)
    - 1. Review permissions: `reviewGeneral` and assigner: `assignGeneral`

>     - 2. Consolidation permission: `reviewAssessmentLevel2` self-assignable

Reviewer: I thought we were going to call this "reviewSenior" or something like that?

Author: This is just to keep the list of permissions short. But for now I would say let's merge like this and change later if desirable (remembering to update here as well)

   - Condition: When Reviewer is_last_level & decision is either CONFORM or NON_CONFORM

6. Generate Review Assignments

   - Action: "generateReviewAssignments"
   - Description: "**CORE-ACTION** Generate next level (in current stage) of review assignments after one reviewer submits a review to next level - when applicable"
   - Condition: Always

7. Change Application Status (for applicant - LOQ)

   - Action: "changeStatus"
   - Description: "**CORE-ACTION** Change application status to CHANGES_REQUIRED once a last level reviewer submits with decision of LIST_OF_QUESTIONS"
   - Condition: Reviwer is_last_level & decision is LIST_OF_QUESTIONS

8. Change Application Outcome (on last stage submission)

   - Action:
   - Description: "**CORE-ACTION** Change application outcome when last level and last stage - usually a final decision - submits a review as CONFORM or NON_CONFORM. Outcome will be APPROVED or REJECTED accordingly"
   - Condition: When Reviewer is_last_stage & is_last_level & decision is either CONFORM or NON_CONFORM

9. Change Application Outcome (on Rejection in early stage)

### Other actions

#### Trigger: ON_APPLICATION_SUBMIT

> - Asynchronous console log

Reviewer: Do we still need this? Probably not in core actions, right?

Author: No - see it's a different section:

## Other actions

1. **TODO** Add example of PDF generation

##### Actions

- Asynchronous console log
  - Action: "consoleLogger"
  - Description: "Example: console log with application serial and template code"
  - Condition: Always

#### Trigger: ON_REVIEW_SUBMIT

1. Change Application Outcome (on Rejection in early stage)
2. Generate new Organisation when APPROVED
3. Add User to new Organisation
4. Grant Permissions to User in Organisation
5. Send Email to User when APPROVED
6. Send Email to User when REJECTED
7. Send Email to User when LIST_OF_QUESTIONS
8. **TODO** Add example of expiry
9. **TODO** Add example of certificate

##### Actions

1. Change Application Outcome (on Rejection in early stage)

   - Action: "changeOutcome"
   - Description: "Example action - Change outcome to REJECT in case of NON_CONFORM decision in first stage of Screening."

>    - Condition: When Reviewer is_last_level & stage is Sreening & decision is NON_CONFORM

Reviewer: Maybe add a comment that the database automatically updates the Status to "COMPLETED" when Outcome is changed from "PENDING", so it's clear why we're not changing status here.

Author: Yes, will do. Thanks
